cudax/include/cuda/experimental/__async/basic_sender.cuh

```diff
@@ -0,0 +1,249 @@
+
+template <class _Completions>
+constexpr bool __has_stopped =
+  !_CUDA_VSTD::is_same_v<__async::completion_signatures<>,
+                         __async::transform_completion_signatures<_Completions,
+                                                                  __async::completion_signatures<>,
+                                                                  __ignore_value_signature,
+                                                                  __ignore_error_signature>>;
+
+template <bool _PotentiallyThrowing, class _Rcvr>
+void set_current_exception_if([[maybe_unused]] _Rcvr& __rcvr) noexcept
+{
+  if constexpr (_PotentiallyThrowing)
+  {
+    __async::set_error(static_cast<_Rcvr&&>(__rcvr), ::std::current_exception());
+  }
+}
+
+// A generic type that holds the data for an async operation, and
+// that provides a `start` method for enqueuing the work.
+template <class _Sndr, class _Data, class _Rcvr>
+struct basic_opstate
+{
+  using __rcvr_t        = basic_receiver<_Data, _Rcvr>;
+  using __completions_t = completion_signatures_of_t<_Sndr, __rcvr_t>;
+  using __traits_t      = __mk_completions<__has_stopped<__completions_t>, _Data, _Rcvr>;
+
+  using completion_signatures =
+    transform_completion_signatures<__completions_t,
+                                    __async::completion_signatures<>, // TODO
+                                    __traits_t::template __set_value_t,
+                                    __traits_t::template __set_error_t,
+                                    typename __traits_t::__set_stopped_t>;
+
+  _CCCL_HOST_DEVICE basic_opstate(_Sndr&& __sndr, _Data __data, _Rcvr __rcvr)
+      : state_{static_cast<_Data&&>(__data), static_cast<_Rcvr&&>(__rcvr)}
+      , __op_(__async::connect(static_cast<_Sndr&&>(__sndr), __rcvr_t{state_}))
+  {}
+
+  _CCCL_HOST_DEVICE _CUDAX_ALWAYS_INLINE void start() noexcept
+  {
+    __async::start(__op_);
+  }
+
+  __state<_Data, _Rcvr> state_;
+  __async::connect_result_t<_Sndr, __rcvr_t> __op_;
+};
+
+template <class _Sndr, class _Rcvr>
+_CCCL_HOST_DEVICE _CUDAX_ALWAYS_INLINE auto __make_opstate(_Sndr __sndr, _Rcvr __rcvr)
+{
+  auto [tag, __data, __child] = static_cast<_Sndr&&>(__sndr);
+  using data_t                = decltype(__data);
+  using child_t               = decltype(__child);
+  return basic_opstate(static_cast<child_t&&>(__child), static_cast<data_t&&>(__data), static_cast<_Rcvr&&>(__rcvr));
+}
+
+template <class _Data, class... _Sndrs>
+_CCCL_HOST_DEVICE _CUDAX_ALWAYS_INLINE auto
+__get_attrs(int, const _Data& __data, const _Sndrs&... __sndrs) noexcept -> decltype(__data.get_attrs(__sndrs...))
+{
+  return __data.get_attrs(__sndrs...);
+}
+
+template <class _Data, class... _Sndrs>
+_CCCL_HOST_DEVICE _CUDAX_ALWAYS_INLINE auto
+__get_attrs(long, const _Data& __data, const _Sndrs&... __sndrs) noexcept -> decltype(__async::get_env(__sndrs...))
+{
+  return __async::get_env(__sndrs...);
+}
+
+template <class _Data, class... _Sndrs>
+struct basic_sender;
```

Comment:
it will be defined eventually. my plan is to specialize `basic_sender` for zero, one, and two children for fewer template instantiations, and to do the `tuple` thing for the fallback case. this is a work in progress. your questions are totally appropriate.